File: ois/ois20/round2/ois11_round2_solutions/machine.cpp
/*
 * This template is valid both in C and in C++,
 * so you can expand it with code from both languages.
 * NOTE: it is recommended to use this even if you don't
 * understand the following code.
 */

#include <assert.h>
#include <stdio.h>
#include <limits>
#include <vector>

using namespace std;

// constraints
#define MAXD 1000

// input data
int C, D, Y, i;
int M[MAXD + 1];
int P[MAXD + 1];

const int INF = numeric_limits<int>::max();

int main() {
  //  uncomment the following lines if you want to read/write from files
  //  freopen("input.txt", "r", stdin);
  //  freopen("output.txt", "w", stdout);

  assert(3 == scanf("%d %d %d", &C, &D, &Y));
  M[0] = 0;
  P[0] = 0;
  // warning! M and P are 1-based
  for (i = 1; i <= D; i++) assert(1 == scanf("%d", &M[i]));
  for (i = 1; i <= D; i++) assert(1 == scanf("%d", &P[i]));

  // calculate cumulative cost for i years
  vector<int> cumulative(D+1);
  for (int i = 1; i <= D; i++) {
      cumulative[i] = cumulative[i-1] + M[i];
  }
  for (int i = 1; i <= D; i++) {
      cumulative[i] += C;
      cumulative[i] -= P[i];
  }

  vector<int> DP(Y+1, INF);
  // only keep useful years
  vector<int> useful;
  DP[0] = 0;
  for (int i = 0; i <= D && i < Y; i++) {
      for (int j = 1; j <= D && i+j <= Y; j++) {
          DP[i+j] = min(DP[i+j], DP[i] + cumulative[j]);
      }

      if (DP[i] == cumulative[i]) {
          // This year is useful
          useful.push_back(i);
      }
  }

  // compute remaining years
  for (int i = D; i < Y; i++) {
      for (int j : useful) {
          if (i+j <= Y) {
              DP[i+j] = min(DP[i+j], DP[i] + cumulative[j]);
          }
      }
  }

  printf("%d\n", DP[Y]);  // print the result
  return 0;
}

File: ois/ois20/round2/ois11_round2_solutions/machine.pas
{ NOTE: it is recommended to use this even if you don't understand the following code. }

{ constraints }
const
    MAXD = 1000;
    MAXY = 1000000;

{ input data }
var
    C, D, Y, i  : longint;
    // Warning! M and P are 1-based
    M, P        : array[1..MAXD] of longint;
    cumulative  : array[1..MAXD] of longint;
    DP : array[0..MAXY] of longint;
    j, k, u : longint;
    useful : array[0..MAXD] of longint;

begin
{
    uncomment the following lines if you want to read/write from files
    assign(input,  'input.txt');  reset(input);
    assign(output, 'output.txt'); rewrite(output);
}

    readln(C, D, Y);
    // Warning! M and P are 1-based
    for i:=1 to D do
        read(M[i]);
    readln();
    for i:=1 to D do
        read(P[i]);
    readln();

    cumulative[1] := M[1];
    for i:=2 to D do
        cumulative[i] := cumulative[i-1] + M[i];

    for i:=1 to D do
    begin
        cumulative[i] := cumulative[i] + C;
        cumulative[i] := cumulative[i] - P[i];
    end;

    DP[0] := 0;
    for i:= 1 to Y do
        DP[i] := 2000000000;

    u := 0;
    for i:= 0 to D do
    begin
        for j:=1 to D do
        begin
            if (i+j <= Y) then
            begin
                if (DP[i+j] >= DP[i] + cumulative[j]) then
                    DP[i+j] := DP[i] + cumulative[j];
            end;
        end;
        if ((i <= Y) and (DP[i] = cumulative[i])) then
        begin
            useful[u] := i;
            u := u+1;
        end;
    end;

    for i:= D to Y do
    begin
        for k:=0 to u-1 do
        begin
            j := useful[k];
            if (i+j <= Y) then
            begin
                if (DP[i+j] >= DP[i] + cumulative[j]) then
                    DP[i+j] := DP[i] + cumulative[j];
            end;
        end;
    end;

    writeln(DP[Y]); { print result }
end.
